### PR TITLE
Remove lr_scheduler requirement in lora_dpo_single_device

### PR DESCRIPTION
**Pull request by thomasjpfan**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Towards https://github.com/pytorch/torchtune/issues/1751

For testing, I initially wanted to `monkeypatch` the `recipes` module, but it was not designed to be imported from tests: https://github.com/pytorch/torchtune/blob/912af640a422d84ce261c100f804490a5c12e8e7/recipes/__init__.py#L7-L8. So for this PR, I added a new integration test that makes a new config without a scheduling and run `tune run` for one epoch.

#### Changelog
What are the changes made in this PR?
* Removes `lr_scheduler` requirement in lora_dpo_single_device

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/lora_dpo_single_device.py [hunk 1]
```diff
@@ -243,7 +243,7 @@ def setup(self, cfg: DictConfig) -> None:
         # Learning rate scheduler can only be set up after number of steps
         # has been computed
         self._lr_scheduler = self._setup_lr_scheduler(
-            cfg_lr_scheduler=cfg.lr_scheduler,
+            cfg_lr_scheduler=cfg.get("lr_scheduler", None),
```
Comment (RdoubleA) [inline recipes/lora_dpo_single_device.py:246]: may be less indirection to just check directly if lr_scheduler exists and set to None here, instead of calling the setup method only to return None:
```
cfg_lr_scheduler = cfg.get("lr_scheduler", None)
self._lr_scheduler = self._setup_lr_scheduler(...) if cfg_lr_scheduler is not None else None
```

File: tests/recipes/test_lora_dpo_single_device.py [hunk 2]
```diff
@@ -191,3 +191,49 @@ def test_save_and_load_merged_weights(self, tmpdir, monkeypatch):
         llama2_model.load_state_dict(sd)
         merged_ckpt_out = llama2_model(inputs)
         torch.testing.assert_close(baseline_out, merged_ckpt_out, rtol=1e-5, atol=1e-5)
+
+    @pytest.mark.integration_test
+    def test_lr_scheduler_optional(self, tmpdir, monkeypatch):
```
Comment (RdoubleA) [inline tests/recipes/test_lora_dpo_single_device.py:196]: this is a great test but it seems expensive just to check that the lr_scheduler is not created. Maybe you could just call the recipe's `setup()` method with a toy model and check that `self._lr_scheduler == None`?

**Review by RdoubleA — COMMENTED** [review]
Thanks for making this update, had a few comments on implementation

File: tests/recipes/test_lora_dpo_single_device.py (hunk 2 above)
Comment (thomasjpfan) [reply tests/recipes/test_lora_dpo_single_device.py:196]: I initially wanted to go with this solution, but I do not think the `recipes` were designed to be imported in tests:

https://github.com/pytorch/torchtune/blob/912af640a422d84ce261c100f804490a5c12e8e7/recipes/__init__.py#L7-L8

Comment (thomasjpfan) [reply tests/recipes/test_lora_dpo_single_device.py:196]: I initially wanted to go with this approach, but I do not think we are allowed to import anything in `recipes` (i.e. `LoRADPORecipeSingleDevice`) from tests:

https://github.com/pytorch/torchtune/blob/912af640a422d84ce261c100f804490a5c12e8e7/recipes/__init__.py#L19-L23

File: recipes/lora_dpo_single_device.py (hunk 1 above)
Comment (felipemello1) [reply recipes/lora_dpo_single_device.py:246]: That makes sense to me, Rafi, but upon second thought, i like the idea of handling everything inside of the setup_lr_scheduler, including the log_info. What do you think?

**Comment by felipemello1** [comment]
Thanks for the pr, @thomasjpfan! I left a couple of comments. After Rafi and Evan reply, i am good with merging it. PS: after the changes for lora_dpo_single_device are done, would you be willing to add it to the other recipes too? Its ok if the answer is no. We can do it as a follow up :)

File: recipes/lora_dpo_single_device.py (hunk 1 above)
Comment (RdoubleA) [reply recipes/lora_dpo_single_device.py:246]: Sure, no strong opinions

**Comment by thomasjpfan** [comment]
> PS: after the changes for lora_dpo_single_device are done, would you be willing to add it to the other recipes too? Its ok if the answer is no. We can do it as a follow up :)

I'll be happy to add it to the other recipes as a followup. I started off with one recipe to see how much test converge is required.

**Comment by felipemello1** [comment]
Just before we approve it, do you mind running it with wandb logger (or any other logger) and capturing the learning rate with/without the scheduler, as a sanity check?

You can append this to your cli command:

```
tune run <...> /
metric_logger=torchtune.training.metric_logging.WandBLogger 
```

**Comment by thomasjpfan** [comment]
With `lr_scheduler._component_=`, the LR changes:

Screenshot 2024-11-15 at 9 28 24 PM

I was not able to get `~lr_scheduler._component_` to remove the config, so I just removed the `lr_scheduler` from the yaml, and the lr stays constant:

Screenshot 2024-11-15 at 9 28 28 PM

**Review by RdoubleA — APPROVED** [review]

**Comment by felipemello1** [comment]
Thank you @thomasjpfan , i created an issue to add to the other recipe. We would love your contribution if you have interest: https://github.com/pytorch/torchtune/issues/2021
